File: core/src/main/java/intrusii/core/service/SubscriptionService.java
package intrusii.core.service;

import intrusii.core.model.Subscription;
import intrusii.core.model.SubscriptionType;

import java.util.List;

public interface SubscriptionService {

    Subscription addSubscription(Subscription subscription);

    void deleteSubscription(Long id);

    Subscription updateSubscription(Subscription subscription);

    List<Subscription> getAllSubscriptions();

    List<Subscription> sortSubscriptionsByType(String isAsc);

    List<Subscription> sortSubscriptionsByDuration(String isAsc);

    List<Subscription> sortSubscriptionsByPrice(String isAsc);

    Subscription getSubscriptionById(Long id);

    List<Subscription> filterSubscriptionByType(SubscriptionType type);

    List<Subscription> filterSubscriptionByDuration(int duration);

    List<Subscription> filterSubscriptionByPrice(float price);
}

File: web/src/main/java/intrusii/web/webapp/src/app/subscriptions/shared/subscription.service.ts
import { Injectable } from '@angular/core';

import {HttpClient} from "@angular/common/http";

import {Subscription} from "./subscription.model";

import {Observable} from "rxjs";

@Injectable({
  providedIn: 'root'
})

export class SubscriptionService {
  private subscriptionsUrl = 'http://localhost:8080/api/subscriptions';

  constructor(private httpClient: HttpClient) { }

  addSubscription(subscription: Subscription): Observable<Subscription> {
    return this.httpClient
      .post<Subscription>(this.subscriptionsUrl, subscription);
  }

  deleteSubscription(id: number): Observable<Subscription> {
    return this.httpClient
      .delete<Subscription>(`${this.subscriptionsUrl}/${id}`);
  }

  updateSubscription(subscription: Subscription): Observable<Subscription> {
    return this.httpClient
      .put<Subscription>(`${this.subscriptionsUrl}/${subscription.id}`, subscription);
  }

  getSubscriptions(): Observable<Subscription[]> {
    return this.httpClient
      .get<Array<Subscription>>(this.subscriptionsUrl);
  }

  sortSubscriptionsByType(isAsc: string): Observable<Subscription[]> {
    return this.httpClient
      .get<Array<Subscription>>(`${this.subscriptionsUrl}/sortByType/${isAsc}`);
  }

  sortSubscriptionsByDuration(isAsc: string): Observable<Subscription[]> {
    return this.httpClient
      .get<Array<Subscription>>(`${this.subscriptionsUrl}/sortByDuration/${isAsc}`);
  }

  sortSubscriptionsByPrice(isAsc: string): Observable<Subscription[]> {
    return this.httpClient
      .get<Array<Subscription>>(`${this.subscriptionsUrl}/sortByPrice/${isAsc}`);
  }

  filterSubscriptionsByType(type: string): Observable<Subscription[]> {
    return this.httpClient
      .get<Array<Subscription>>(`${this.subscriptionsUrl}/filterByType/${type}`);
  }

  filterSubscriptionsByDuration(duration: string): Observable<Subscription[]> {
    return this.httpClient
      .get<Array<Subscription>>(`${this.subscriptionsUrl}/filterByDuration/${duration}`);
  }

  filterSubscriptionsByPrice(price: string): Observable<Subscription[]> {
    return this.httpClient
      .get<Array<Subscription>>(`${this.subscriptionsUrl}/filterByPrice/${price}.`);
  }
}
